### PR TITLE
CTAP2.1

### PR DESCRIPTION
**Pull request by AdamVe**
Adds [CTAP2.1](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html)

Changes:
- implemented support for [enterpriseAttestation](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#makecred-enterpriseattestation)
- added CTAP2.1 members of [authenticatorGetInfo](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#authenticatorGetInfo) response
- added support for [permissions](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#perms-param-permissions)/[permissionsRpId](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#authenticatorclientpin-rpid) in [authenticatorClientPin](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#authnrClientPin-cmd-dfn) 
- added support for following authenticatorClientPin subcommands: [getPinUvAuthTokenUsingPinWithPermissions](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#getPinUvAuthTokenUsingPinWithPermissions), [getPinUvAuthTokenUsingUvWithPermissions](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#getPinUvAuthTokenUsingUvWithPermissions) and getUVRetries
- easily accessible powerCycleState value through calling [getPINRetries](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#gettingPINRetries). 
- added support for [PIN/UV Auth Protocol Two](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#pinProto2)
- added support for [authenticatorCredentialManagement](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#authenticatorCredentialManagement).
- added support for [authenticatorConfig](https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#authenticatorConfig) command on the ctap level 
- testing framework has been updated to support parameters and conditions for easier test execution.

### CODE REVIEW

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/PinUvAuthProtocolV2.java [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+        } catch (NoSuchAlgorithmException | InvalidKeySpecException | InvalidKeyException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    @Override
+    public byte[] encrypt(byte[] key, byte[] demPlaintext) {
+        try {
+            Cipher cipher = Cipher.getInstance(CIPHER_TRANSFORMATION);
+            cipher.init(Cipher.ENCRYPT_MODE, new SecretKeySpec(key, CIPHER_ALG), new IvParameterSpec(IV));
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/PinUvAuthProtocolV2.java]: this has been fixed in following commit

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/PinUvAuthProtocolV2.java [hunk 2]
```diff
@@ -0,0 +1,160 @@
...
+                 IllegalBlockSizeException e) {
+            throw new IllegalStateException(e);
+        }
+    }
+
+    @Override
+    public byte[] decrypt(byte[] key, byte[] ciphertext) {
+        try {
+            Cipher cipher = Cipher.getInstance(CIPHER_TRANSFORMATION);
+            cipher.init(Cipher.DECRYPT_MODE, new SecretKeySpec(key, CIPHER_ALG), new IvParameterSpec(IV));
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/PinUvAuthProtocolV2.java]: This has been fixed in [1087183](https://github.com/Yubico/yubikit-android/pull/105/commits/1087183fc43b16d93d071c490f7867530c37e4ef)

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+        }
+        return hmacDigest(salt, ikm);
+    }
+
+    byte[] expand(byte[] prk) throws NoSuchAlgorithmException, InvalidKeyException {
+        byte[] t = new byte[0];
+        byte[] okm = new byte[0];
+        byte i = 0;
+        while (okm.length < length) {
+            i += 1;
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java]: Fixed in [e571d67](https://github.com/Yubico/yubikit-android/pull/105/commits/e571d67acf236199921f552ccb78bef2efb3f419)

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java [hunk 4]
```diff
@@ -0,0 +1,86 @@
...
+ * @see <a href="https://datatracker.ietf.org/doc/html/rfc5869">rfc5869</a>
+ */
+public class Hkdf {
+
+    private final int length;
+    private final byte[] salt;
+    private final byte[] info;
+    private final Mac mac;
+
+    public Hkdf(String algo, @Nullable byte[] salt, byte[] info, int length) throws NoSuchAlgorithmException {
```
Comment (dainnilsson) [inline fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java:40]: I find it a little hard to follow where some of these methods use arguments that shadow the instance members. For example, `extract` takes `salt` even though it could use `this.salt`. On the other hand, `expand` uses `(this.)length` and `(this.)info`.
I'm also not sure why the constructor accepts a null `salt`, as we always explicitly pass it.

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java [hunk 5]
```diff
@@ -0,0 +1,86 @@
...
+import javax.annotation.Nullable;
+import javax.crypto.Mac;
+import javax.crypto.spec.SecretKeySpec;
+
+/**
+ * Implements HMAC-based Extract-and-Expand Key Derivation Function (HKDF)
+ *
+ * @see <a href="https://datatracker.ietf.org/doc/html/rfc5869">rfc5869</a>
+ */
+public class Hkdf {
```
Comment (dainnilsson) [inline fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java:33]: I would prefer this class to be package-private to avoid making it part of the public API.

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java (hunk 4 above)
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/Hkdf.java]: Updated in [18961b5](https://github.com/Yubico/yubikit-android/pull/105/commits/18961b55f93a80af3f8a02221a993b3567aaa687) 

`extract` and `expand` now also have the same parameters as in the rfc. See [2.2 Step 1: Extract](https://datatracker.ietf.org/doc/html/rfc5869#section-2.2) and [2.3 Step2: Expand](https://datatracker.ietf.org/doc/html/rfc5869#section-2.3)

File: fido/src/main/java/com/yubico/yubikit/fido/client/BasicWebAuthnClient.java [hunk 6]
```diff
@@ -67,10 +70,10 @@
...
 @SuppressWarnings("unused")
 public class BasicWebAuthnClient implements Closeable {
-    private static final String OPTION_CLIENT_PIN = "clientPin";
-    private static final String OPTION_CREDENTIAL_MANAGEMENT = "credentialMgmtPreview";
-    private static final String OPTION_USER_VERIFICATION = "uv";
-    private static final String OPTION_RESIDENT_KEY = "rk";
+    public static final String OPTION_CLIENT_PIN = "clientPin";
+    public static final String OPTION_USER_VERIFICATION = "uv";
+    public static final String OPTION_RESIDENT_KEY = "rk";
+    public static final String OPTION_EP = "ep";
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/client/BasicWebAuthnClient.java]: No reason. Reverted the change and the fields are private now.

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Config.java [hunk 7]
```diff
@@ -0,0 +1,200 @@
...
+        parameters.put(PARAM_NEW_MIN_PIN_LENGTH, minPinLength);
+        parameters.put(PARAM_MIN_PIN_LENGTH_RPIDS, rpIds);
+        parameters.put(PARAM_FORCE_CHANGE_PIN, forceChangePin != null && forceChangePin);
+
+        call(CMD_SET_MIN_PIN_LENGTH, parameters);
+        Logger.info(logger, "Minimum PIN length set");
+    }
+
+    /**
+     * Set the minimum PIN length allowed when setting/changing the PIN.
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/Config.java]: good eyes! fixed this.

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/CredentialManagement.java [hunk 8]
```diff
@@ -66,13 +71,34 @@ public class CredentialManagement {
...
         this.ctap = ctap;
+        Ctap2Session.InfoData ctapInfo = ctap.getCachedInfo();
+        Map<String, ?> options = ctapInfo.getOptions();
+
+        if (Boolean.TRUE.equals(options.get("credMgmt"))) {
+            cmd = Ctap2Session.CMD_CREDENTIAL_MANAGEMENT;
+        } else if (ctapInfo.getVersions().contains("FIDO_2_1_PRE")
+                &&
+                Boolean.TRUE.equals(options.get("credentialMgmtPreview"))) {
+            cmd = Ctap2Session.CMD_CREDENTIAL_MANAGEMENT_PRE;
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/CredentialManagement.java]: I have updated this in https://github.com/Yubico/yubikit-android/pull/105/commits/916c619ea098c9b191b2cb329fbaf0a2a168875b

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java [hunk 9]
```diff
@@ -334,65 +462,175 @@ public void close() throws IOException {
...
+                @Nullable Boolean forcePINChange,
+                @Nullable Integer minPINLength,
+                @Nullable Integer firmwareVersion,
+                @Nullable Integer maxCredBlobLength,
+                @Nullable Integer maxRPIDsForSetMinPINLength,
+                @Nullable Integer preferredPlatformUvAttempts,
+                @Nullable Integer uvModality,
+                @Nullable Map<String, Object> certifications,
+                @Nullable Integer remainingDiscoverableCredentials,
+                @Nullable List<Integer> vendorPrototypeConfigCommands) {
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java]: I reviewed these and updated in [5c90614](https://github.com/Yubico/yubikit-android/pull/105/commits/5c906140d9732b903237aa2018668f09d4b1729d)

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java [hunk 10]
```diff
@@ -644,16 +636,32 @@ private static InfoData fromData(Map<Integer, ?> data) {
...
                             ? (List<String>) data.get(RESULT_TRANSPORTS)
                             : Collections.emptyList(),
-                    (List<PublicKeyCredentialParameters>) data.get(RESULT_ALGORITHMS),
-                    (Integer) data.get(RESULT_MAX_SERIALIZED_LARGE_BLOB_ARRAY),
-                    (Boolean) data.get(RESULT_FORCE_PIN_CHANGE),
-                    (Integer) data.get(RESULT_MIN_PIN_LENGTH),
+                    data.containsKey(RESULT_ALGORITHMS)
+                            ? (List<PublicKeyCredentialParameters>) data.get(RESULT_ALGORITHMS)
+                            : Collections.emptyList(),
+                    data.containsKey(RESULT_MAX_SERIALIZED_LARGE_BLOB_ARRAY)
```
Comment (dainnilsson) [inline fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java:642]: For maxSerializedLargeBlobArray and maxCredBlobLength the spec states that if the key is not present the feature is not supported. 1024 and 32 are minimum values IF the feature is supported. Because of this I think it would be more appropriate to use 0 for both default values.

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java [hunk 11]
```diff
@@ -99,36 +111,47 @@ public static void create(YubiKeyDevice device, Callback<Result<Ctap2Session, Ex
     }
 
     public Ctap2Session(SmartCardConnection connection)
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java:104]: fixed

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java [hunk 12]
```diff
@@ -172,6 +195,29 @@ byte[] sendCbor(byte[] data, @Nullable CommandState state) throws IOException {
...
+    }
+
+    /**
+     * Get information about authenticator support of credential manager commands.
+     * @return true if the authenticator supports credential manager or credential manager preview
+     * commands are supported.
+     * @see <a href="https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#ref-for-getinfo-credmgmt">credMgmt option</a>
+     * @see <a href="https://fidoalliance.org/specs/fido-v2.1-ps-20210615/fido-client-to-authenticator-protocol-v2.1-ps-errata-20220621.html#ref-for-getinfo-credentialmgmtpreview">credentialMgmtPreview option</a>
+     */
+    public boolean isCredentialManagerSupported() {
```
[the remark this answers is not in the document]
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java]: I have updated this in https://github.com/Yubico/yubikit-android/pull/105/commits/98c361453dee2fd9b10d1aa6b241d09b86437c80.

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java (hunk 10 above)
Comment (AdamVe) [reply fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java:620]: fixed

**Review by dainnilsson — APPROVED** [review]

File: core/src/main/java/com/yubico/yubikit/core/fido/FidoProtocol.java [hunk 13]
```diff
@@ -35,14 +35,16 @@ public class FidoProtocol implements Closeable {
 
     public static final byte TYPE_INIT = (byte) 0x80;
 
-    private static final byte CMD_PING = TYPE_INIT | 0x01;
-    private static final byte CMD_APDU = TYPE_INIT | 0x03;
-    private static final byte CMD_INIT = TYPE_INIT | 0x06;
-    private static final byte CMD_WINK = TYPE_INIT | 0x08;
-    private static final byte CMD_CANCEL = TYPE_INIT | 0x11;
+    public static final byte CTAPHID_PING = TYPE_INIT | 0x01;
```
Comment (AdamVe) [inline core/src/main/java/com/yubico/yubikit/core/fido/FidoProtocol.java:38]: make private

File: fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java [hunk 14]
```diff
@@ -229,17 +231,17 @@ public CredentialData makeCredential(
             @Nullable Map<String, ?> extensions,
             @Nullable Map<String, ?> options,
             @Nullable byte[] pinUvAuthParam,
-            int pinUvAuthProtocol,
+            @Nullable Integer pinUvAuthProtocol,
+            @Nullable Integer enterpriseAttestation,
```
Comment (AdamVe) [inline fido/src/main/java/com/yubico/yubikit/fido/ctap/Ctap2Session.java:235]: no enterprise attestation parameter

File: fido/src/main/java/com/yubico/yubikit/fido/client/BasicWebAuthnClient.java [hunk 15]
```diff
@@ -96,8 +101,20 @@ public BasicWebAuthnClient(Ctap2Session session) throws IOException, CommandExce
 
         Boolean clientPin = (Boolean) options.get(OPTION_CLIENT_PIN);
         pinSupported = clientPin != null;
-        if (pinSupported && info.getPinUvAuthProtocols().contains(PinUvAuthProtocolV1.VERSION)) {
-            this.clientPin = new ClientPin(ctap, new PinUvAuthProtocolV1());
+
+        final List<Integer> pinUvAuthProtocols = info.getPinUvAuthProtocols();
+        if (pinUvAuthProtocols.size() > 0) {
```
Comment (AdamVe) [inline fido/src/main/java/com/yubico/yubikit/fido/client/BasicWebAuthnClient.java:106]: rewrite this to select any supported protocol version
